### PR TITLE
feat(php8): match expression

### PR DESCRIPTION
**Pull request by cseufert**
Added new AST objects for Match and  MatchEntry

### CODE REVIEW

**Comment by czosel** [comment]
PHP-Parser is parsing 
```php
match(trye) {
  0,1,2,3 => run(),
  default => null,
};
```
as
```
              expr: Expr_Match(
                cond: Expr_ConstFetch(
                    name: Name(
                        parts: array(
                            0: trye
                        )
                    )
                )
                arms: array(
                    0: MatchArm(
                        conds: array(
                            0: Scalar_LNumber(
                                value: 0
                            )
                            1: Scalar_LNumber(
                                value: 1
                            )
                            2: Scalar_LNumber(
                                value: 2
                            )
                            3: Scalar_LNumber(
                                value: 3
                            )
                        )
                        body: Expr_FuncCall(
                            name: Name(
                                parts: array(
                                    0: run
                                )
                            )
                            args: array(
                            )
                        )
                    )
                    1: MatchArm(
                        conds: null
                        body: Expr_ConstFetch(
                            name: Name(
                                parts: array(
                                    0: null
                                )
                            )
                        )
                    )
                )
            )
```
Notable differences:

- `arms` instead of `body` 
- `MatchArm` instead of `MatchEntry`
- `conds` instead of `keys`
- `default` is not parsed as a special keyword, but as `conds: null`

The PHP [documentation](https://www.php.net/manual/en/control-structures.match.php) also speaks of arms, it might be a good idea to follow this? Also, parsing `default` as `conds: null` seems like a simple solution to me - what do you think @cseufert ?

**Comment by cseufert** [comment]
Thanks for the guidance on this, btw, what PHP-Parser project are you referencing?

I have updated the match statement, as per the code above.

**Comment by czosel** [comment]
AST looks good! I'll try to take a look at the code later today. Here is the parser I was refering to: https://github.com/nikic/PHP-Parser

File: src/ast.js [hunk 1]
```diff
@@ -486,6 +486,7 @@ AST.prototype.checkNodes = function () {
   require("./ast/declaration"),
   require("./ast/declare"),
   require("./ast/declaredirective"),
+  require("./ast/defaultkeyword"),
```
Comment (czosel) [inline src/ast.js:489]: not needed anymore

File: src/ast/defaultkeyword.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * @extends {Node}
+ */
+module.exports = Node.extends(KIND, function DefaultKeyword(
+  raw,
+  docs,
+  location
+) {
+  Node.apply(this, [KIND, docs, location]);
+  this.raw = raw;
+});
```
Comment (czosel) [inline src/ast/defaultkeyword.js:23]: not needed anymore

**Review by czosel — COMMENTED** [review]
Just some cleanup, otherwise looks good to me :+1:

File: src/ast/defaultkeyword.js (hunk 2 above)
Comment (cseufert) [reply src/ast/defaultkeyword.js]: Fixed

File: src/ast.js (hunk 1 above)
Comment (cseufert) [reply src/ast.js]: Fixed

**Comment by cseufert** [comment]
@czosel I think this is now pretty much complete, however I havent checked for all syntax corner cases in tests, but the normal ones are in there.

**Review by czosel — APPROVED** [review]
:+1:
